hyperspy/io_plugins/emd.py

```diff
@@ -0,0 +1,374 @@
+    user: dictionary
+        Dictionary which contains user related metadata.
+    microscope: dictionary
+        Dictionary which contains microscope related metadata.
+    sample: dictionary
+        Dictionary which contains sample related metadata.
+    comments: dictionary
+        Dictionary which contains additional commentary metadata.
+
+    '''
+
+    _log = logging.getLogger(__name__)
+
+    def __init__(self, signals={}, user={}, microscope={}, sample={}, comments={}):
+        self._log.debug('Calling __init__')
+        # Make sure some default keys are present in user:
+        for key in ['name', 'institution', 'department', 'email']:
+            if key not in user:
+                user[key] = ''
+        self.user = user
+        # Make sure some default keys are present in microscope:
+        for key in ['name', 'voltage']:
+            if key not in microscope:
+                microscope[key] = ''
+        self.microscope = microscope
+        # Make sure some default keys are present in sample:
+        for key in ['material', 'preparation']:
+            if key not in sample:
+                sample[key] = ''
+        self.sample = sample
+        # Add comments:
+        self.comments = comments
+        # Make sure the signals are added properly to signals:
+        self.signals = {}
+        for name, signal in signals.iteritems():
+            self.add_signal(signal, name)
+
+    def __getitem__(self, key):
+        # This is for accessing the raw data easily. For the signals use emd.signals[key]!
+        return self.signals[key].data
+
+    def _write_signal_to_group(self, data_group, signal):
+        self._log.debug('Calling _write_signal_to_group')
+        # Save data:
+
+        dataset = data_group.create_group(signal.metadata.General.title)
+        dataset['data'] = signal.data
+        # Iterate over all dimensions:
+        for i in range(len(signal.data.shape)):
+            key = 'dim{}'.format(i+1)
+            offset = signal.axes_manager[i].offset
+            scale = signal.axes_manager[i].scale
+            dim = dataset.create_dataset(key, data=[offset, offset+scale])
+            name = signal.axes_manager[i].name
+            from traits.trait_base import _Undefined
+            if type(name) is _Undefined:
+                name = ''
+            dim.attrs['name'] = name
+            units = signal.axes_manager[i].units
+            if type(units) is _Undefined:
+                units = ''
+            else:
+                units = '[{}]'.format('_'.join(list(units)))
+            dim.attrs['units'] = units
+        # Write metadata:
+        dataset.attrs['emd_group_type'] = 1
+        for key, value in signal.metadata.Signal:
+            dataset.attrs[key] = value
+
+    def _read_signal_from_group(self, name, group):
+        self._log.debug('Calling _read_signal_from_group')
+        import hyperspy.api as hp
+        # Extract essential data:
+        data = group.get('data')[...]

```

Comment:
You can just pass the `h5py.Dataset` as-is. It should ideally avoid loading it into memory unless we really need to. It's easy to get the numpy array later anyway